R/staggregate.R

```diff
@@ -0,0 +1,601 @@
+
+
+  # Data.table of weights
+  weights_dt <- data.table::as.data.table(overlay_weights)
+
+  # Extent of area weights with slight buffer to make sure all cells are included
+  min_x <- min(weights_dt$x) - 0.5
+  max_x <- max(weights_dt$x) + 0.5
+  min_y <- min(weights_dt$y) - 0.5
+  max_y <- max(weights_dt$y) + 0.5
+
+  weights_ext <- raster::extent(min_x, max_x, min_y, max_y)
+
+  ## Load climate data
+  ## -----------------------------------------------
+
+
+  # Immediately crop to weights extent
+  clim_raster <- raster::crop(raster::stack(data), weights_ext)
+
+  # Get layer names (dates)
+  all_layers <- names(clim_raster)
+  layer_names <- all_layers[seq(1, length(all_layers), 24)] # Keep every 24th layer name (1 per day)
+  layer_names <- paste0('year_', substring(layer_names, 2,5), 'month_', substring(layer_names, 7,8), '_day_', substring(layer_names, 10,11)) # Extract month/day for each layer
+
+  ## Aggregate to grid-day level
+  ## -----------------------------------------------
+
+  if(!daily_agg %in% c('average', 'sum')){
+
+    stop(crayon::red("Daily aggregation must be 'average' or 'sum'"))
+  }
+
+  ## Average
+  if(daily_agg == 'average'){
+
+    message(crayon::green("Averaging hourly values to get daily values"))
+
+    # Average over each set of 24 layers
+    indices<-rep(1:(raster::nlayers(clim_raster)/24),each=24)
+    clim_daily <- raster::stackApply(clim_raster, indices = indices, fun=mean) #Stack of 365/366 layers
+  }
+
+  ## Sum
+  if(daily_agg == 'sum'){
+
+    message(crayon::green("Summing hourly values to daily values"))
+
+    # Sum over each set of 24 layers
+    indices<-rep(1:(raster::nlayers(clim_raster)/24),each=24)
+    clim_daily <- raster::stackApply(clim_raster, indices = indices, fun=sum) #Stack of 365/366 layers
+  }
+
+
+
+  # Return a list containing, in order, daily aggregated climate data as a raster brick, and the layer_names created.
+  return(list(clim_daily, layer_names))
+}
+
+
+# Function to merge with geoweights and aggregate by polygon
+polygon_aggregation <- function(clim_dt, weights_dt, list_names, time_agg){
+
+  ## Merge weights with climate raster
+  ## -----------------------------------------------
+
+  # Set key column in the climate data.table
+  keycols = c("x", "y")
+  data.table::setkeyv(clim_dt, keycols)
+
+  # Keyed merge on the x/y column
+  merged_dt <- clim_dt[weights_dt, allow.cartesian = TRUE] #cols: x, y, date, value cols 1:k, poly_id, w_area, weight (if weights = T)
+
+  ## Multiply weights x climate value (all 1:k values); aggregate by month and polygon
+  ## -----------------------------------------------
+
+  # Multiply by secondary weights if weights = TRUE (already normalized by polygon area)
+  # Otherwise multiply by just area weights
+  if("weight" %in% names(merged_dt)){
+    merged_dt[, (list_names) := lapply(list_names, function(x) {get(x) * weight})]
+  } else {
+    merged_dt[, (list_names) := lapply(list_names, function(x) {get(x) * w_area})]
+  }
+
```

Comment:
I think instead of having to add the date transformation to every transformation function, we can just add it here once. Right now the `date` column of `clim_dt` should have one date for every day and they should still be in the format `Xyear.month.day.hour.min.sec`. We can add these two lines to remove the `X` and then convert the column to a date. 
```
clim_dt[, date := stringr::str_remove(date, 'X')]
clim_dt[, date := lubridate::as_date(date)]
```
Or if we want to be a bit more flexible we can remove the first character of the string: 
```
clim_dt[, date := stringr::str_sub(date, 2, -1)] 
clim_dt[, date := lubridate::as_date(date)]
```